generic/shared/src/main/scala/io/circe/generic/JsonCodecMacros.scala

```diff
@@ -0,0 +1,50 @@
+package io.circe.generic
+
+import io.circe.{ Decoder, Encoder }
+import macrocompat.bundle
+import scala.language.experimental.macros
+import scala.reflect.macros.blackbox
+
+class JsonCodec extends scala.annotation.StaticAnnotation {
+  def macroTransform(annottees: Any*): Any = macro JsonCodecMacros.jsonCodecAnnotationMacro
+}
+
+@bundle
+private[generic] class JsonCodecMacros(val c: blackbox.Context) {
+  import c.universe._
+
+  def jsonCodecAnnotationMacro(annottees: Tree*): Tree = annottees match {
+    case List(clsDef: ClassDef) if clsDef.mods hasFlag Flag.CASE =>
+      q"""
+       $clsDef
+       object ${clsDef.name.toTermName} {
+         ..${codec(clsDef.name)}
+       }
+       """
+    case List(
+      clsDef: ClassDef,
+      q"object $objName extends { ..$objEarlyDefs } with ..$objParents { $objSelf => ..$objDefs }"
+    ) if clsDef.mods hasFlag Flag.CASE =>
+      q"""
+       $clsDef
+       object $objName extends { ..$objEarlyDefs} with ..$objParents { $objSelf =>
+         ..${codec(clsDef.name)}
+         ..$objDefs
+       }
+       """
+    case _ => c.abort(c.enclosingPosition, "Invalid annotation target: must be a case class")
+  }
+
+  private[this] val DecoderClass = symbolOf[Decoder[_]]
+  private[this] val EncoderClass = symbolOf[Encoder[_]]
+  private[this] val semiautoObj = symbolOf[semiauto.type].asClass.module
+
+  private[this] def codec(tpname: TypeName): List[Tree] = {
+    val decodeNme = TermName("decode" + tpname.decodedName)
```

Comment:
I'm don't think using `fresh` is a good idea here, because it's not stable between compiles.

Are you happy to leave this as is for now and revisit it in the future?